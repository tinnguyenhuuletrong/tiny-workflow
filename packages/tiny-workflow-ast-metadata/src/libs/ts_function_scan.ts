import {
  MethodDeclaration,
  Project,
  PropertyAccessExpression,
  SyntaxKind,
} from "ts-morph";

export function scan_expression_call(
  methodDeclaration: MethodDeclaration,
  methodName: string
) {
  const callExpressions = methodDeclaration.getDescendantsOfKind(
    SyntaxKind.CallExpression
  );

  // Filter the call expressions to find those like `this.withAction`
  const withActionCalls = callExpressions.filter((callExpr) => {
    const expression = callExpr.getExpression();

    // Ensure that the expression is a PropertyAccessExpression
    if (expression.getKind() === SyntaxKind.PropertyAccessExpression) {
      const propertyAccess = expression as PropertyAccessExpression;

      // Check if the expression matches `this.withAction`
      return (
        // propertyAccess.getExpression().getText() === "this" &&
        propertyAccess.getName() === methodName
      );
    }

    return false;
  });

  return withActionCalls;
}
