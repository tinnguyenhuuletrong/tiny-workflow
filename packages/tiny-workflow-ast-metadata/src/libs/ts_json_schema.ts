import { type Type, type Node, Project } from "ts-morph";

export function convertTypeToSchema(type: Type, node: Node): any {
  const schema: any = {};

  if (type.isString()) {
    schema.type = "string";
  } else if (type.isNumber()) {
    schema.type = "number";
  } else if (type.isBoolean()) {
    schema.type = "boolean";
  } else if (type.isArray()) {
    schema.type = "array";
    schema.items = convertTypeToSchema(type.getArrayElementTypeOrThrow(), node);
  } else if (type.isObject()) {
    schema.type = "object";
    schema.properties = {};
    const properties = type.getProperties();
    for (const prop of properties) {
      const propType = prop.getTypeAtLocation(node);
      schema.properties[prop.getName()] = convertTypeToSchema(propType, node);
    }
  } else if (type.isUnion()) {
    const unionVal = type.getUnionTypes();
    schema.type = "string";
    schema.enum = unionVal.map((itm) => itm.getText().slice(1, -1));
  } else {
    schema.type = "unknown";
  }

  return schema;
}
