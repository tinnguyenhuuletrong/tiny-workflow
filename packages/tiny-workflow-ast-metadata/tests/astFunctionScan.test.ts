import { test, expect, mock } from "bun:test";

import { scan_expression_call } from "../src/libs/ts_function_scan";
import { Project } from "ts-morph";

test("ast-jsonSchema-test", async () => {
  const code = `
class A {
  async fun_1() {
    this.withAction('k1', async () => {})
    this.withNoAction('k2', async () => {})
    this.withAction('k3', async () => {})
  }
}
  `;

  const project = new Project({});
  const sourceFile = project.createSourceFile("code.ts", code);

  // Get the class declaration by name
  const classDeclaration = sourceFile.getClassOrThrow("A");

  // Get the method (function) by name within the class
  const methodDeclaration = classDeclaration.getMethodOrThrow("fun_1");

  // Scan the method body for call expressions
  const withActionCalls = scan_expression_call(methodDeclaration, "withAction");

  const allActionKeys = withActionCalls.map((itm) =>
    itm.getArguments()?.[0].getText().slice(1, -1)
  );
  expect(allActionKeys).toMatchSnapshot();
});
