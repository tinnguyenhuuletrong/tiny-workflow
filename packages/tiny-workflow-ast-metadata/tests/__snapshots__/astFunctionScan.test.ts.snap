// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`ast-jsonSchema-test 1`] = `
[
  "k1",
  "k3",
]
`;
